File: thesis-api/main/services/question_service/edit_distance_service.py
from unidecode import unidecode


class EditDistanceService:
    def delta2(self, a: str, b: str) -> float:
        if a.lower() == b.lower():
            return 0
        if unidecode(a) == unidecode(b):
            return 0.5
        return 1

    def edit_distance(self, x: str, y: str, delta=None) -> float:
        if delta is None:
            delta = self.delta2
        edit_table = [[None] * (len(y) + 1) for _ in range(len(x) + 1)]
        for i in range(1, len(x) + 1):
            edit_table[i][0] = (i, i - 1, 0)
        for j in range(1, len(y) + 1):
            edit_table[0][j] = (j, 0, j - 1)
        edit_table[0][0] = (0, 0, 0)
        for i in range(len(x)):
            k = i + 1
            for j in range(len(y)):
                l = j + 1
                edit_table[k][l] = (edit_table[k - 1][l][0] + 1, k - 1, l)
                if edit_table[k][l - 1][0] + 1 < edit_table[k][l][0]:
                    edit_table[k][l] = (edit_table[k][l - 1][0] + 1, k, l - 1)
                if edit_table[k][l][0] > edit_table[k - 1][l - 1][0] + delta(x[i], y[j]):
                    edit_table[k][l] = (edit_table[k - 1][l - 1][0] + delta(x[i], y[j]), k - 1, l - 1)
        return edit_table[-1][-1][0]

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/EditDistanceService.java
package pl.edu.agh.quizzesthesis.business;

import me.xuender.unidecode.Unidecode;
import org.springframework.stereotype.Service;

@Service
public class EditDistanceService {

    public float editDistance(String a, String b) {
        float[][][] editTable = new float[a.length() + 1][b.length() + 1][3];
        for (int i = 1; i <= a.length(); i++) {
            editTable[i][0] = new float[]{i, i - 1, 0};
        }
        for (int j = 1; j <= b.length(); j++) {
            editTable[0][j] = new float[]{j, 0, j - 1};
        }
        editTable[0][0] = new float[]{0, 0, 0};

        for (int i = 0; i < a.length(); i++) {
            int k = i + 1;
            for (int j = 0; j < b.length(); j++) {
                int l = j + 1;
                editTable[k][l] = new float[]{editTable[k - 1][l][0] + 1, k - 1, l};
                if (editTable[k][l - 1][0] + 1 < editTable[k][l][0]) {
                    editTable[k][l] = new float[]{editTable[k][l - 1][0] + 1, k, l - 1};
                }
                if (editTable[k][l][0] > editTable[k - 1][l - 1][0] + delta2(a.charAt(i), b.charAt(j))) {
                    editTable[k][l] = new float[]{editTable[k - 1][l - 1][0] + delta2(a.charAt(i), b.charAt(j)), k - 1, l - 1};
                }
            }
        }
        return editTable[a.length()][b.length()][0];
    }

    private float delta2(char a, char b) {
        if (a == b) {
            return 0.0f;
        } else if (Unidecode.decode(String.valueOf(a)).equals(Unidecode.decode(String.valueOf(b)))) {
            return 0.5f;
        } else {
            return 1.0f;
        }
    }
}
